src/data/matrix/kronecker_product_bialgebra.lean

```diff
@@ -0,0 +1,198 @@
+@[simp] def kronecker_map (f : α → β → γ) (A : matrix l m α) (B : matrix n p β) :
+  matrix (l × n) (m × p) γ
+| i j := f (A i.1 j.1) (B i.2 j.2)
+
+lemma kronecker_map_zero_left [has_zero α] [has_zero γ] (f : α → β → γ) (hf : ∀ b, f 0 b = 0)
+  (B : matrix n p β) :
+  kronecker_map f (0 : matrix l m α) B = 0:=
+ext $ λ i j,hf _
+
+lemma kronecker_map_zero_right [has_zero β] [has_zero γ] (f : α → β → γ) (hf : ∀ a, f a 0 = 0)
+  (A : matrix l m α) :
+  kronecker_map f A (0 : matrix n p β) = 0 :=
+ext $ λ i j, hf _
+
+lemma kronecker_map_add_left [has_add α] [has_add γ] (f : α → β → γ)
+  (hf : ∀ a₁ a₂ b, f (a₁ + a₂) b = f a₁ b + f a₂ b)
+  (A₁ A₂ : matrix l m α) (B : matrix n p β) :
+  kronecker_map f (A₁ + A₂) B = kronecker_map f A₁ B + kronecker_map f A₂ B :=
+ext $ λ i j, hf _ _ _
+
+lemma kronecker_map_add_right [has_add β] [has_add γ] (f : α → β → γ)
+  (hf : ∀ a b₁ b₂, f a (b₁ + b₂) = f a b₁ + f a b₂)
+  (A : matrix l m α) (B₁ B₂ : matrix n p β) :
+  kronecker_map f A (B₁ + B₂) = kronecker_map f A B₁ + kronecker_map f A B₂ :=
+ext $ λ i j, hf _ _ _
+
+lemma kronecker_map_smul_left [has_scalar R α] [has_scalar R γ] (f : α → β → γ)
+  (hf : ∀ (r : R) a b, f (r • a) b = r • f a b) (r : R)
+  (A : matrix l m α) (B : matrix n p β) :
+  kronecker_map f (r • A) B = r • kronecker_map f A B :=
+ext $ λ i j, hf _ _ _
+
+lemma kronecker_map_smul_right [has_scalar R β] [has_scalar R γ] (f : α → β → γ)
+  (hf : ∀ (r : R) a b, f a (r • b) = r • f a b) (r : R)
+  (A : matrix l m α) (B : matrix n p β) :
+  kronecker_map f A (r • B) = r • kronecker_map f A B :=
+ext $ λ i j, hf _ _ _
+
+/-- When `f` is bilinear then `matrix.kronecker_map f` is also bilinear. -/
+@[simps]
+def kronecker_map_linear [comm_semiring R]
+  [add_comm_monoid α] [add_comm_monoid β] [add_comm_monoid γ]
+  [module R α] [module R β] [module R γ]
+  (f : α →ₗ[R] β →ₗ[R] γ) :
+  matrix l m α →ₗ[R] matrix n p β →ₗ[R] matrix (l × n) (m × p) γ :=
+linear_map.mk₂ R
+  (kronecker_map (λ r s, f r s))
+  (kronecker_map_add_left _ $ f.map_add₂)
+  (kronecker_map_smul_left _ $ f.map_smul₂)
+  (kronecker_map_add_right _ $ λ a, (f a).map_add)
+  (kronecker_map_smul_right _ $ λ r a, (f a).map_smul r)
+
+end
+
+
+variables {α R S β : Type*} [comm_semiring α] [comm_semiring R] [comm_semiring S] [comm_semiring β]
+variables [algebra α R] [algebra α S] [algebra α β] [algebra R β] [algebra S β]
+variables {l m n p l' m' n' p' : Type*}
+variables [fintype l] [fintype m] [fintype n] [fintype p]
+variables [fintype l'] [fintype m'] [fintype n'] [fintype p']
+
+variables (α β)
+
+-- TODO: move this
+def algebra.biprod [is_scalar_tower α R β] [is_scalar_tower α S β] :
+  R →ₗ[α] S →ₗ[α] β :=
+((algebra.lmul α β).to_linear_map.compl₂ ((algebra.linear_map S β).restrict_scalars α)).comp
+  ((algebra.linear_map R β).restrict_scalars α)
+
+@[simp]
+lemma algebra.biprod_apply [is_scalar_tower α R β] [is_scalar_tower α S β] (r : R) (s : S) :
+  algebra.biprod α β r s = algebra_map R β r * algebra_map S β s := rfl
```

Comment:
Just to check; this is a generalization that is actually useful to something you want to or think someone might want to do, right? That is, you do actually care about

* `matrix l m R →ₗ[α] matrix n p S →ₗ[α] matrix (l × n) (m × p) β`

And not just

* `matrix l m R →ₗ[α] matrix n p R→ₗ[α] matrix (l × n) (m × p) R`

Reply:
Well, let's say that to me the more natural setting is the first one (actually with
```math
β = R \otimes_{\alpha} S
```
but I have no direct application of this so far (and I grab the chance to ask you how do you type Math/Unicode symbols in this chat).

Reply:
I agree that the tensor product setting is more natural. If they're not actually what you want, I'd be tempted to ditch the biproducts and just  have the homogenous version; you can spell `kronecker_biprod A B` as `kronecker_prod (A.map $ algebra_map _ _) (B.map $ algebra_map _ _)`, and if you don't actually care about it right now that extra verbosity isn't relevant.

What are your thoughts on this @jcommelin?

Reply:
(regarding unicode symbols, I do a mixture of copying from your diff, and using the android keyboard user dictionary I posted in zulip)

Reply:
>  you can spell `kronecker_biprod A B` as `kronecker_prod (A.map $ algebra_map _ _) (B.map $ algebra_map _ _)`

#8468 provides `linear_map.map_matrix`, so this can be kept bundled as something like
```
(kronecker_prod.comp (algebra.linear_map _ _).map_matrix).compl2 (algebra.linear_map _ _).map_matrix)
```